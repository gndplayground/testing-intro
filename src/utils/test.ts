import renderer from "react-test-renderer";
import React from "react";

export function rendererToJson(a: React.ReactElement) {
  const component = renderer.create(a);
  const tree = component.toJSON();
  return { component, tree };
}
