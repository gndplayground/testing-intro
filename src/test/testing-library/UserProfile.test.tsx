import React from "react";
import { render, fireEvent, act } from "@testing-library/react";
import UserProfile from "./UserProfile";
import { rendererToJson } from "../../utils/test";

describe("<UserProfile />", () => {
  it("Should render default", () => {
    const { tree } = rendererToJson(<UserProfile />);
    expect(tree).toMatchSnapshot();
  });

  it("Should display the profile", async () => {
    const userProfile = {
      name: "Jon Doe",
    };

    const fetchMock = jest.spyOn(global, "fetch");

    fetchMock.mockImplementation(() => {
      return Promise.resolve({
        json: () => Promise.resolve(userProfile),
        // json: () => {
        //   return new Promise((resolve) => {
        //     setTimeout(() => {
        //       resolve(userProfile);
        //     }, 500);
        //   });
        // },
      }) as any;
    });

    const { getByTestId, getByText, container } = render(<UserProfile />);

    const button = getByTestId("request-profile");

    await act(async () => {
      fireEvent.click(button);
    });

    // --env=jsdom-fourteen
    // await waitFor(() => {
    //   expect(button).not.toBeInTheDocument();
    // });

    expect(button).not.toBeInTheDocument();

    expect(fetchMock).toBeCalled();

    getByText(`Hello ${userProfile.name}`);

    expect(container.firstChild).toMatchSnapshot();
  });
});
