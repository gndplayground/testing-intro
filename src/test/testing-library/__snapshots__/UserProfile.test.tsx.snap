// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserProfile /> Should display the profile 1`] = `
<section>
  <h2>
    User Profile
  </h2>
  <p>
    Hello 
    Jon Doe
  </p>
</section>
`;

exports[`<UserProfile /> Should render default 1`] = `
<section>
  <h2>
    User Profile
  </h2>
  <button
    data-testid="request-profile"
    onClick={[Function]}
  >
    Request profile
  </button>
</section>
`;
